src/ImpactEvaluator.sol

```diff
@@ -88,45 +87,50 @@ contract ImpactEvaluator is AccessControl {
         Round storage round = openRounds[roundIndex];
         require(round.exists, "Open round does not exist");
 
-        for (uint i = 0; i < addresses.length; i++) {
-            round.addresses.push(addresses[i]);
-            round.scores.push(scores[i]);
-        }
+        uint sumOfScores = validateScores(scores, round.totalScores);
+        reward(addresses, scores);
+        round.totalScores += sumOfScores;
 
-        if (allScoresSubmitted(round.scores)) {
-            reward(round.addresses, round.scores);
-            cleanUpRound(round, roundIndex);
+        if (round.totalScores == MAX_SCORE) {
+            delete openRounds[roundIndex];
         }
     }
 
-    function cleanUpRound(Round storage round, uint roundIndex) private {
-        round.end = 0;
-        delete round.addresses;
-        delete round.scores;
-        round.exists = false;
-        delete openRounds[roundIndex];
-    }
-
-    function allScoresSubmitted(
-        uint64[] memory scores
-    ) private pure returns (bool) {
-        uint totalScore = 0;
+    function validateScores(
+        uint64[] memory scores,
+        uint scoresAlreadySubmitted
+    ) private pure returns (uint) {
+        uint64 sum = 0;
         for (uint i = 0; i < scores.length; i++) {
-            totalScore += scores[i];
+            sum += scores[i];
         }
-        return totalScore >= MAX_SCORE;
+        require(sum <= MAX_SCORE, "Sum of scores too big");
+        require(
+            sum + scoresAlreadySubmitted <= MAX_SCORE,
+            "Sum of scores including historic too big"
+        );
+        return sum;
     }
 
     function reward(
         address payable[] memory addresses,
         uint64[] memory scores
     ) private {
-        validateScores(scores);
-        require(address(this).balance >= roundReward, "Not enough funds");
+        uint totalAmount = 0;
+        uint[] memory amounts = new uint[](addresses.length);
+
         for (uint i = 0; i < addresses.length; i++) {
-            address payable addr = addresses[i];
             uint score = scores[i];
             uint256 amount = (score * roundReward) / MAX_SCORE;
+            amounts[i] = amount;
+            totalAmount += amount;
+        }
+
+        require(address(this).balance >= totalAmount, "Not enough funds");
```

Comment:
Im wondering if the logic for the contracts having enough funds should be in the actual contract itself? I think it should be on the evaluation service to check because the contract balance is publicly available.  

In my opinion, the IE contract should just distribute whatever funds are available.

Reply:
I have added this here so that we don't even start looping over addresses and paying out, if we know we don't have enough funds to complete it. This is to prevent some participants from getting funds while others won't. Or wdyt?

Reply:
Yeah seems like a good idea to check for some funds in the balance before distributing rewards. 

I wonder if we can make this check more efficient by just comparing:
`require(address(this).balance >= roundReward, "Not enough funds");` and we avoid the loop entirely. 

In the previous checks, we only trigger the reward function if the sum of the scores is equal to the `MAX_SCORE` so the loop here feels redundant.

Reply:
This is the purpose of the PR though: We know pay rewards for each batch of evaluations submitted, effectively batching the rewards as well. Prior to this, the participant count would be limited by gas fees, now we can evaluate/reward in as many batches as we want.

This means however that for each batch we need to check balance by verifying we have enough for this batch, and not the whole round.